keys.hh

```diff
@@ -784,10 +784,18 @@ auto format_pk(const WithSchemaWrapper& pk, FormatContext& ctx) {
                                                                      components.end()))) |
                       boost::adaptors::transformed([](const auto& type_and_component) {
                           auto& [type, component] = type_and_component;
-                          auto key = type->to_string(to_bytes(component));
+                          auto bs = to_bytes(component);
+                          auto key = type->to_string(bs);
                           if (!utils::utf8::validate((const uint8_t *) key.data(), key.size())) {
                               return sstring("<non-utf8-key>");
                           }
+                          if (!std::all_of(key.begin(), key.end(), [] (auto ch) { return std::isprint(ch); })) {
+                            std::stringstream s;
+                            for (const auto& b : bs) {
+                                fmt::print(s, "\\x{:02x}", b);
+                            }
+                            return sstring(s.str());
+                          }
```

Comment:
i'd suggest use something like
```c++
                      boost::adaptors::transformed([](const auto& type_and_component) -> std::string {
                          auto& [type, component] = type_and_component;
                          auto bytes = to_bytes(component);
                          auto str = type->to_string(bytes);
                          if (std::all_of(str.begin(), str.end(), [] (auto ch) { return std::isprint(ch); })) {
                              return str;
                          } else {
                              return fmt::format("\\{:1\\}", bytes);
                          }
                      }),
```

for following reasons:

1. `isprint()` is much stricter than `utf8::validate()`, as the former only recognize a subset of ASCII, while the letter covers a much larger code points, even including multi-bytes characters. so since we have a fallback, the non-utf8 branch can be dropped.
2. no need to use `std::stringstream` if we already have a formatter for bytes. see https://github.com/scylladb/scylladb/blob/aba31ad06c60c103ad80fbfa1d35376315f6ae41/bytes.hh#L56
3. prefer using positive condition for better readability.

Reply:
> i'd suggest use something like
> 
> ```c++
>                       boost::adaptors::transformed([](const auto& type_and_component) -> std::string {
>                           auto& [type, component] = type_and_component;
>                           auto bytes = to_bytes(component);
>                           auto str = type->to_string(bytes);
>                           if (std::all_of(str.begin(), str.end(), [] (auto ch) { return std::isprint(ch); })) {
>                               return str;
>                           } else {
>                               return fmt::format("\\{:1\\}", bytes);
>                           }
>                       }),
> ```
> 
> for following reasons:
> 
> 1. `isprint()` is much stricter than `utf8::validate()`, as the former only recognize a subset of ASCII, while the letter covers a much larger code points, even including multi-bytes characters. so since we have a fallback, the non-utf8 branch can be dropped.

But couldn't a valid utf8 string contain non-printable character (0x7f in particular)?

> 2. no need to use `std::stringstream` if we already have a formatter for bytes. see https://github.com/scylladb/scylladb/blob/aba31ad06c60c103ad80fbfa1d35376315f6ae41/bytes.hh#L56

a. I pushed a new version using `sstring(initialized_later, size)`

b. regarding fmt_hex, the hex content is already present in the next field in the log printout, repeating it here doesn't add anything.

I even contemplated if the \x{:02x} style is required at all, or just print something like `<non-printable-key>`, as the hex value is printed right after it...

> 3. prefer using positive condition for better readability.

Can use `any_of`, but then the predicate would be negative :)

Reply:
ah I see above what you meant by positive condition...

Reply:
BTW, what does this format string mean: `"\\{:1\\}"`?
I couldn't figure that out from https://fmt.dev/latest/syntax.html#grammar-token-sf-format_spec

Reply:
sorry, i meant to put:

```c++
            boost::adaptors::transformed([](const auto& type_and_component) -> std::string {
                          auto& [type, component] = type_and_component;
                          auto bytes = to_bytes(component);
                          auto str = type->to_string(bytes);
                          if (std::all_of(str.begin(), str.end(), [] (auto ch) { return std::isprint(ch); })) {
                              return str;
                          } else {
                              return fmt::format("\\{:1\\}", fmt_hex(bytes));
                          }
                      }),
```

Reply:
> > i'd suggest use something like
> > ```c++
> >                       boost::adaptors::transformed([](const auto& type_and_component) -> std::string {
> >                           auto& [type, component] = type_and_component;
> >                           auto bytes = to_bytes(component);
> >                           auto str = type->to_string(bytes);
> >                           if (std::all_of(str.begin(), str.end(), [] (auto ch) { return std::isprint(ch); })) {
> >                               return str;
> >                           } else {
> >                               return fmt::format("\\{:1\\}", bytes);
> >                           }
> >                       }),
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > for following reasons:
> > 
> > 1. `isprint()` is much stricter than `utf8::validate()`, as the former only recognize a subset of ASCII, while the letter covers a much larger code points, even including multi-bytes characters. so since we have a fallback, the non-utf8 branch can be dropped.
> 
> But couldn't a valid utf8 string contain non-printable character (0x7f in particular)?

yes. that's why we don't need to filter out the non-utf8 string anymore. as we are able to print all of them now.

> 
> > 2. no need to use `std::stringstream` if we already have a formatter for bytes. see https://github.com/scylladb/scylladb/blob/aba31ad06c60c103ad80fbfa1d35376315f6ae41/bytes.hh#L56
> 
> a. I pushed a new version using `sstring(initialized_later, size)`
> 
> b. regarding fmt_hex, the hex content is already present in the next field in the log printout, repeating it here doesn't add anything.

repeated? i don't follow you. my formatter does not repeat anything.

> 
> I even contemplated if the \x{:02x} style is required at all, or just print something like `<non-printable-key>`, as the hex value is printed right after it...
> 
> > 3. prefer using positive condition for better readability.
> 
> Can use `any_of`, but then the predicate would be negative :)

Reply:
> BTW, what does this format string mean: `"\\{:1\\}"`? I couldn't figure that out from https://fmt.dev/latest/syntax.html#grammar-token-sf-format_spec

please refer to the link in the first comment.